Support Multi-View Images

danieldresser-ie opened the pull request:
This is a big PR, and I'm still doing some cleanup ( mostly on ImageWriter, which can be implemented more cleanly after ContextQuery, but also a few other TODOs, and some needed testing ), but I think it's pretty ready for review in most areas ( maybe hold off on ImageWriter ).

johnhaddon reviewed (commented):
Thanks Daniel!

This is a monster of a PR! I don't mind saying that I've found it a bit overwhelming to review - I've been on it the majority of the day, and still haven't finished. Posting what I have so far, as I don't trust GitHub not to lose the pending comments overnight.

I don't think I commented on the ImageReader changes, but I did spend time going through them and didn't see any problems. I do wonder if there's a way of refactoring the File constructor so it's a little easier to follow - perhaps branching at a high level into dedicated single-part and multi-part initialisers? Perhaps there's little immediate value in that now though, since we have it working and we have good test coverage.

One thing I don't think I mentioned in the inline comments is that we're a bit inconsistent about whether we're calling things `viewName` or just `view`. I think I'd probably said `viewName` originally and then in a couple of spots (like `CopyViews.views`) you've gone for just `view`. After seeing it, I definitely prefer the brevity of the latter, so perhaps we should just use that consistently thoughout? With a lot of the earlier naming I'd pedantically used `Name` suffixes because technically it's not the thing itself, but in practice I think this is just annoying, and more recent nodes have avoided it.

I'll take a look at the remaining commits tomorrow...

Cheers...
John

johnhaddon commented on include/GafferImage/ImagePlug.h:
```diff
@@ -166,6 +166,36 @@ class GAFFERIMAGE_API ImagePlug : public Gaffer::ValuePlug
+		/// it is in validViews, or validViews is empty and there is no view set ),
+		/// then ViewCheckScope does nothing, and isPresent will return true.
+		///
+		/// If there is a view set, but validViews is empty, then ViewCheckScope will
+		/// create a scope that remove view from the context, and isValid will return
+		/// true ( a viewless input should affect every view ).
```
Are we sure this is the right behaviour? Hypothetical example :

- I am merging a viewless image over a stereo one.
- I want to tweak something in that viewless image for one of the destination views.
- If we delete the `image:view` context variable, then I need to use CreateViews to convert my image to stereo first, then tweak it downstream of the CreateViews.
- If we just leave `image:view` in place, I can just go ahead and use it directly in an expression or spreadsheet.

In other words, does it make sense for the default view to be treated as "all views" rather than "no view"?

johnhaddon replied:
I think what I'm saying here is probably a red herring, and it doesn't sit well with the other stuff I'm saying about having a `default` view. I'll leave it here though, in case it sparks any thoughts...

danieldresser-ie replied:
Yeah, I think this is a red herring - I think we may want to treat the default view as "all views" when we use it, but I think we still need to clean up the view context before evaluating plugs ( rather than source nodes needing to accept queries of views which don't exist ).

johnhaddon reviewed (commented):
OK, day two review complete! I can truly appreciate the scale of what you've been working on now @danieldresser-ie, so thanks for taking it on! I've done my best to digest it all, but I think we'd both benefit from breaking it into chunks that we can get over the line individually from here. Before we do that though, I'd like to take one step back, and make a proposal for your consideration...

### Viewless images considered harmful

We've currently got two distinct states an image can be in :

- Viewless : `viewNames` is empty, and we require that `image:viewName` not be specified in the context.
- Viewed : `viewNames` is non-empty, and we require that `image:viewName` be present in the context, with one of those values.

I think we arrived at this approach for two reasons :

- OpenEXR has the concept of viewless images, and we want to be able to round-trip arbitrary EXR files.
- We were attracted to the viewless state because it is an exact match for what we were doing previously.

But after digesting the PR, it's not sitting well for me :

- It forces the user to swallow another concept, the "viewless" image, on top of the "multiview" concept we're trying to introduce. In our conversations on the topic, I've noticed a real awkwardness in trying to express this distinction and to reason about it. I don't think it's an easy concept to "get" or to communicate.
- It forces all client code into an awkward `if views.size() {} else {}` pattern.
- Every time I see `emptyViewNames` anywhere, I'm having to say to myself "I know your instinct is that this means there is no image, but actually there is, there is a _viewless_ image". It's not clicking for me.
- OpenEXR has the concept of viewless channels in an otherwise multi-view file. It'd be nice if we could represent this too, but I don't think it's something we can bolt on cleanly afterwards. We'd either need a separate flag, or a special value in viewNames, say `[ "", "left", "right" ]`. But then we're left with more special cases - a "magic" transition back to `[]` (a viewless image) if we delete `left right` from `[ "", "left", "right" ]`. And how do we delete `""`?

I propose the following remedy :

- In Gaffer, everything belongs to a view. The default image state for things like FlatImageSource is `viewNames = [ "default" ]`.
- The ImageWriter and ImageReader  map the `default` view to and from OpenEXR's "channels not in a view".
- The `image:view` context variable must be provided when querying all image properties other than `views`.

This is far easier to explain, easier to code for, and can better represent the full set of possibilities in an EXR file. I'm sure I'm missing some details that you can point out, but I'm feeling pretty hopeful that this is the better approach...

danieldresser-ie commented on include/GafferImage/ImageNode.h:
```diff
@@ -121,6 +122,7 @@ class GAFFERIMAGE_API ImageNode : public Gaffer::ComputeNode
 		void compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const override;
 		/// Compute methods for the individual children of outPlug() - these must be implemented by derived classes, or
 		/// an input connection must be made to the plug, so that the method is not called.
+		virtual IECore::ConstStringVectorDataPtr computeViewNames( const Gaffer::Context *context, const ImagePlug *parent ) const;
```
Hmm, good question.  I guess I probably did just imitate channelNames.  The argument that it would make sense to switch both at the same time is reasonable - but on the other side, it would be quicker to switch now when there's no compatibility concerns, and if we did viewNames this way, and it works smoothly, it would be a good demonstration that it would work well to switch channelNames ( which would be a lot more work, because of compatibility and channelNames being used in more places )

johnhaddon replied:
Summary of separate conversation : we'll get this PR across the line using strings for view name, and then do a quick follow up to use InternedString.

danieldresser-ie commented on python/GafferImageTest/ImageReaderTest.py:
```diff
@@ -561,5 +573,317 @@ def testWithChannelTestImage( self ):
+						# assertAlmostEqual doesn't use significant digits.  Should really put a proper
+						# version of assertAlmostEqual in GafferTest.TestCase ( we could handle imath types
+						# as well ), and also handle nans properly
+						if not ( stats["average"].getValue()[i] != stats["average"].getValue()[i] and e[3][i] != e[3][i] or abs( e[3][i] ) > 1e20 ):
```
I'm using an extremely imprecise tolerance of `places = 4`, because specifying a number of places is a terrible way of setting tolerance, if the values can range widely ( some of the test images contain large values ).  A more precise and meaningful test would be that values are within 6 significant digits

johnhaddon replied:
It was the conditional guarding the `assertAlmostEqual()` that I wasn't following...

danieldresser-ie replied:
The EXR examples try to exercise corner cases, so they include very large numbers ( which would never pass unless we raise the tolerance so much it made the other tests meaningless, because of tolerance being specified badly ), and NaNs ( assertAlmostEqual fails on Nans, and throws if you try to assert Nan == Nan ) - the conditional skips these cases where  assertAlmostEqual doesn't work

danieldresser-ie commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -128,6 +129,17 @@ std::string channelNameFromEXR( std::string view, std::string part, std::string
 		// and is not affected by anything else.  The only special case we need to handle is removing the view
 		// from the name in the case of single part multiview files.
+		
+		if( view.size() )
```
Oh, that's true - when useHeuristics is false, I should only enable that test in the case of single part multiview.

danieldresser-ie commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -258,53 +269,170 @@ class File
+
+					if( viewNames.size() > 0 && viewName == "" )
+					{
+						// Channels without views in multiview files are technically allowed by the spec.
+						// We could support this fairly easily, but John worries it's confusing, and it's a weird
+						// special case anyway, so we're not going to support it unless someone asks for it.
```
Sounds about right, though now it sounds like we're going to switch to using a default view, in which case we will probably support these.

danieldresser-ie commented on src/GafferImage/OpenImageIOReader.cpp:
```diff
@@ -326,116 +454,139 @@ class File
 			// Do the actual read of data
+			Box2i exrTargetRegion = flopDisplayWindow( targetRegion, view.m_imageSpec.full_y, view.m_imageSpec.full_height );
 			std::vector<float> fileData;
 			DeepData fileDeepData;
-			Box2i fileDataRegion;
-			const int nchannels = readRegion( tileBatchIndex.z, targetRegion, fileData, fileDeepData, fileDataRegion );
+			Box2i exrDataRegion;
+
+			const int nchannels = readRegion( tileBatchIndex.z, exrTargetRegion, fileData, fileDeepData, exrDataRegion );
+
+			Box2i fileDataRegion = flopDisplayWindow( exrDataRegion, view.m_imageSpec.full_y, view.m_imageSpec.full_height );
```
The logic hasn't changed, I've just pulled it outside of readRegion - previously readRegion did the flopping itself, based on a global display window.  Now that the display window depends on the view, it felt easier to have readRegion work solely in EXR coordinates, and move the flopping outside to where we're dealing with views.  Note to myself:  assuming we're happy with it this way, I need to update this comment which is now inaccurate:  `// Convert target region in Gaffer into a region of the file that we are trying to read`

danieldresser-ie commented on src/GafferImage/CollectImages.cpp:
```diff
@@ -198,6 +200,68 @@ void CollectImages::affects( const Gaffer::Plug *input, AffectedPlugsContainer &
+		const std::vector< std::string > &layerViewNames = inPlug()->viewNamesPlug()->getValue()->readable();
+		if( outViewNames != inPlug()->viewNamesPlug()->getValue()->readable() )
+		{
+			// Requiring all contexts to have matching view names is quite restrictive, but is the simplest.
```
I don't see any reason to completely disallow CollectLayers for multiview images - considering a simple example, like loading aovs from different files into layers of the same image, there's no reason why you might not want to merge beauty and albedo AOVs for both eyes of a stereo image.  In that example, requiring both beauty and albedo to have matching views is reasonable ( maybe having some sort of fallback for non-matching views would be even better?  But requiring matching is better than nothing ).

johnhaddon replied:
So I'd have a stereo file (left and right views in one file) with the beauty, and a stereo file with the albedo, and I'd be collecting them? Sounds good. Do you think a rename to CollectLayers is reasonable, to help disambiguate against a CollectViews? (not for this PR, but for future work).

danieldresser-ie replied:
Yeah, rename to CollectLayers is reasonable, as is adding CollectViews.  Both for future.

johnhaddon replied:
I'm leaving this one open for now as I think it's worth considering doing the rename as an immediate followup to this PR.

danieldresser-ie commented on src/GafferImage/CreateViews.cpp:
```diff
@@ -0,0 +1,196 @@
+	addChild( s );
+
+	s->setup( new ImagePlug() );
+	s->selectorPlug()->setValue( "${" + ImagePlug::viewNameContextName.string() + "}" );
```
Hmm, that toggle would do the trick here.  The issue would be switches with integer indices, or even a named switched where the name was set by an expression instead of a string substitution.

danieldresser-ie replied:
Still seems like a good idea, but now irrelevant to this PR, since for the new magic we don't tamper with the context vairable here.

danieldresser-ie commented on src/GafferImage/CopyViews.cpp:
```diff
@@ -0,0 +1,303 @@
+size_t CopyViews::g_firstPlugIndex = 0;
+
+CopyViews::CopyViews( const std::string &name )
+	:	ImageProcessor( name, /* minInputs = */ 1 )
```
Oh, I think I was just imitating one of the other Copy nodes - but I think it is reasonable to think that you could want to copy from multiple images, because this is how we implement merging views?

johnhaddon replied:
Yeah, I think we do want an array here. I suppose according to the guidelines in the wiki, it should arguably be a `source` array plug, separate from the single `in` plug. The reasoning being that the logic for the array is different - everything is taken from `in`, but only the matching views are taken from `source`. I'm not sure how much value this has though - maybe for ImageProcessors the single array `in` is preferable? It's not like we have a separate array in `Merge`, just because the operation doesn't apply to the first image...

danieldresser-ie commented on include/GafferImage/ImagePlug.h:
```diff
@@ -166,6 +166,36 @@ class GAFFERIMAGE_API ImagePlug : public Gaffer::ValuePlug
+			bool isPresent();
+			const Gaffer::Context *context();
+		private:
+			std::unique_ptr<Gaffer::Context::EditableScope> m_editableScope;
```
The downside is paying for the size of an EditableScope instead of just a pointer as part of ViewCheckScope in the common case where this scope doesn't get used.  But I'm OK with switching to optional if you prefer.

johnhaddon replied:
Let's see how my ViewScope proposal below pans out, since that would intrinsically _be_ an EditableScope anyway.

danieldresser-ie commented on include/GafferImage/ImagePlug.h:
```diff
@@ -166,6 +166,36 @@ class GAFFERIMAGE_API ImagePlug : public Gaffer::ValuePlug
+		/// If there is no view set, but validViews is non-empty, an exception will be
+		/// thrown.  In order to use a multi view input, views must be declared on the
+		/// output.
+		struct GAFFERIMAGE_API ViewCheckScope
```
I'll need to try implementing it, but sounds reasonable.

danieldresser-ie commented:
OK, I don't think I'm particularly opposed to the default view proposal - it does make the code simpler.

Main question:  does the "default" view actually act as a "default", in the sense of "if you ask for a view that doesn't exist, you get the default instead"?  The code would be even simpler if we didn't treat the default as special at all, and error messages would be simpler ( Just `"left" doesn't exist"` rather than `"left" doesn't exist and there is no default set` ), but it does seem like having it act as a default would be more convenient sometimes.  But if it does act as a default, we need to define where this happens - where we would want it is basically wherever we use ViewCheckScope currently.  But then we need to document that `ImagePlug.channelNames( "left" )` will not look for the default view if left isn't found ... is that weird?

The one other special case I can think of - an exr with a view literally named "default".  It's a little weird that we would fail to round trip that?  We use a double underscore prefix for internal things with special meaning elsewhere, but the confusion of having both "__default" and "default" seems worse than failing to round trip in a special case.

johnhaddon commented on src/GafferImage/FormatQuery.cpp:
```diff
@@ -67,6 +67,24 @@ setV2PlugComponentValue( T const& parent, Gaffer::NumericPlug< typename T::Value
+	ViewScope( const Gaffer::Context *context, std::string &view, const std::vector< std::string > &views )
+		: EditableScope( context )
+	{
+		if( view.size() && views.size() )
+		{
+			if( std::find( views.begin(), views.end(), view ) == views.end() )
+			{
+				throw IECore::Exception( "ImageStats : View does not exist \"" + view + "\"" );
+			}
+			set( GafferImage::ImagePlug::viewNameContextName, &view );
+		}
```
One other potential upside of the `default` view could be that `view == ""` in FormatQuery could mean "don't modify the context", so that by default FormatQuery operates on the view provided by the downstream context. Or do we think folks should need to enter `${image:view}` explicitly for that? We should review the other query nodes for precedent, but I _think_ the scene ones work this way...

johnhaddon replied:
Leaving this open, since I think it warrants a bit of discussion. At the very least, I think we'd need to improve the empty value in the `FormatQuery.view` UI?

danieldresser-ie replied:
Hmm, yes, I had forgotten about the custom UI when I took your suggestion for "view == "" means don't modify context".  Should I add some sort of metadata like `viewPlugValueWidget:allowUseCurrentContext`, which adds a menu option called "Use Current Context", with a value of ""?

johnhaddon replied:
Yeah, that sounds good.

johnhaddon commented:
> But if it does act as a default, we need to define where this happens - where we would want it is basically wherever we use ViewCheckScope currently. But then we need to document that ImagePlug.channelNames( "left" ) will not look for the default view if left isn't found ... is that weird?

It definitely seems to require a bit more explaining, which I always take as a bit of a warning sign. I took a look at a certain popular compositing package to see if there was any precedent there to guide us. But as far as I could tell, that required the set of views to be defined globally for the script, and then the nodes seemed hardwired to only those views. I don't think that's a good fit for Gaffer, where we want views to be determined dynamically by context, and we want to be able to create views in the graph, not globally.

Let's see, what are our options for being more helpful, if we decide that the core functionality should be that the `default` view doesn't have magic powers?

1. Add a `views` plug to FlatImageSource, so you can easily specify what views should be generated at source.
2. Follow the pattern established by the `image:defaultFormat` context variable, so that FlatImageSource et al make views defined by an `image:defaultViews` context variable (unless overridden by their own `views` plug). This provides global control without departing from our principles of everything being dynamic based on context.
3. A PromoteDefaultView node that copies the default view to match the views of another node. This seems like it would have its uses, but be too fiddly to rely on as a general solution.
4. A `doMagicWithDefaultViews` toggle on multi-input ImageProcessors, so the magic is opt-in. Is turn-on-and-offable magic better than always-on magic?
5. Leftfield idea. Don't remove `image:view` from the context when querying `ImagePlug.views`. Then the default for `FlatImageSource.views` could be `${image:view}`, falling back to `default` if its not defined. Bit sketchy...

I'm pretty convinced that 1) is a worthwhile addition in its own right, regardless of anything else. And 2) seems to fit reasonably into our existing conventions, and would work pretty well when pasting nodes from a mono-view script into a stereo-view one. Where I don't think either help is in the building of networks for processing arbitrary numbers of views. How would I make a RenderStats node that adds a text overlay of the render stats from the metadata, that can work with any number of views? Only 3 and 4 would help there, but it'd be fiddly and annoying having to opt in to that.

I don't have a strong opinion on this yet...

danieldresser-ie commented:
1) Does seem reasonable.  2) I guess is consistent, though it doesn't really seem like any sort of solution to the core problem - I think we do need to make it easy to match a data stream in the network.

Your example of the render stats node seems like a pretty good example - the magic doesn't really help with that ( with or without a doMagic toggle ), since the magic would work by requesting the default view in the merge - not giving the upstream text the opportunity to check which view to read metadata for.  Usually, we want to get rid of context variables whenever possible for performance reasons, but with the view driving everything, it probably is better to make sure the view is in the context while evaluating the upstream nodes?  5) would do this, but I feel is way too confusing.  It feels to me like what is needed is a convenient way to plug the viewNames from an imagePlug into the views plug on a source ... it's a shame that one is a vector of strings ( or InternedStrings ), and the other is a space separated string ( actually, should we just support connecting a vector of strings to a string as meaning "convert to space-separated?  We do seem to use that convention a lot ).  However we do this, we need to do it ourselves in Text, since text has an input to match the views of, so your render stats example should just work ( but hopefully however we implement this internally is easy enough that it would also be easy for someone setting up a FlatImageSource to do manually ).

johnhaddon commented:
> Usually, we want to get rid of context variables whenever possible for performance reasons, but with the view driving everything, it probably is better to make sure the view is in the context while evaluating the upstream nodes?

Yeah, I don't think we would ever want to delete the view from the context implicitly. I suppose we might when performing a scene query from an ImageNode, but I think that's fairly rare, and it's just a couple of new contexts, not an explosion.

> 5) would do this, but I feel is way too confusing.

Yep, it's nonsense that `viewNames` would give you the values you can use for `${image:view}`, but would itself be sensitive to that variable. A more reasonable alternative might be a different definition for the magic behaviour of the `default` view :

```
for input in inputs :
    if context["image:viewName"] not in input.viewNames() and "default" not in input.viewNames() :
         continue
    # Never change the context. If the upstream node has the "default" view, that means it will happily output
    # that for _any_ value of `image:view`.
    evaluateInput()
```

I think this would actually work already without needing to modify any of the FlatImageSources. They don't care one bit what view they're evaluated in, and they're not required to check (that is the responsibility of the client). But by leaving the view in the context, we leave it open for their inputs (expressions, spreadsheets, image queries etc) to operate using `${image:view}`. With this in place, I think I might even prefer to omit `FlatImageProcessor.views`.

I'm not sure this is a definitive argument for choosing magic, but I think it would be my preference for the form of magic to consider. Actually, I think I might be fairly keen on it...

> actually, should we just support connecting a vector of strings to a string as meaning "convert to space-separated? We do seem to use that convention a lot

I think that might be reasonable - I actually wanted this the other day, so I could drive `Set.name` with `Spreadsheet.enabledRowNames`. I still think we'd end up with a `ViewQuery` node to do the wiring though, since `ImageNode.out.viewNames` isn't really accessible via the UI. One counterargument would be how we treat InternedStringVectorData in `Context::substitute()` - as a `/slash/separated/path`. I don't think that's a strong argument though - we could/should do something more principled where it only applied to `${scene:path}`.

danieldresser-ie commented on python/GafferImageTest/ImageReaderTest.py:
```diff
@@ -561,5 +573,317 @@ def testWithChannelTestImage( self ):
+
+			with IECore.CapturingMessageHandler() as mh :
+				self.assertEqual( list( multiPartReader["out"].viewNames() ), [ "right", "left" ] )
+				# Test all the channels which are in views, we currently ignore channels not in views
+				# The order is a little bit weird : EXR sorts alphabetically, and then OIIO sorts RGBA correctly.
+				# This matches our expectations:
```
Yeah, I think that would totally makes sense, but if we're trying to limit changes at the moment, it probably doesn't need to be done now.

danieldresser-ie commented on python/GafferImageUI/ViewPlugValueWidget.py:
```diff
@@ -0,0 +1,116 @@
+
+		return None
+
+	def __imagePlugDirtied( self, plug ):
+		if plug == self.__firstInputImagePlug()["viewNames"]:
+			self._updateFromPlug()
```
How should we react to this?  Do you we change the menu label somehow: "`left (not found)`"?  Should it be different for a view that isn't found when "default" is present, and the image can be evaluated for the non-existent view, vs having no default, so the image actually can't be evaluated?

johnhaddon replied:
Yes, that sounds about right. Maybe this?

- View exists : `left`
- View doesn't exist and no default view : `left (invalid)`
- View doesn't exist and default view : `left (default)`

danieldresser-ie commented on python/GafferImageUI/ViewPlugValueWidget.py:
```diff
@@ -0,0 +1,116 @@
+			value = self.getPlug().getValue()
+
+		if len( self._views() ):
+			self.__menuButton.setText( value )
```
I think I am going to debate this.  Do we have other examples like SelectView where a user must type the name of something to select it, but the name displayed in the UI has a capitalization that will fail to work if they type it?  ( Especially when someone could conceivably create a file with two views named "Left" and "left" )

The closest comparison would be image layer names, and we don't switch the case of those.

johnhaddon replied:
Fair point - my eyes will just have to get used to it.

danieldresser-ie commented on src/GafferImage/Merge.cpp:
```diff
@@ -669,10 +671,16 @@ void Merge::hashDataWindow( const GafferImage::ImagePlug *output, const Gaffer::
 	for( ImagePlug::Iterator it( inPlugs() ); !it.done(); ++it )
 	{
-		if( (*it)->getInput<ValuePlug>() )
+		// TODO - it's technically a bit wasteful to be running this on the first iteration - we don't
+		// need to check viewNames on the first iteration, since the output viewNames are identical to
+		// the first iteration viewNames, so the view is guaranteed to be valid, unless the downstream
+		// node has a bug
+		ImagePlug::ViewCheckScope viewCheckScope( context, (*it)->viewNames()->readable() );
```
With the "new magic" now, the merge node is never responsible for modifying the viewName in the Context, it only needs to check whether the input node can support the current viewName.  For checking this, I've added a function currently named ImageAlgo::viewIsValid

danieldresser-ie commented:
OK, I think this is actually ready for a preliminary re-review, with some massive caveats.  The areas I have not addressed are the 4 new nodes:  CreateViews, SelectView, DeleteViews, CopyViews.  These nodes all need tests, and also some fixes identifiied in your review, and they need to be re-examined now that we've switched to the new magic - I think it will probably make more sense to rereview them after squashing.  ( Oh, and I should probably add a couple quick tests for FormatQuery, ImageStats, and ImageSampler working with views )

What makes sense to look at now is anything that isn't one of those 4 nodes, which you want to review before squashing.  There are also some miscellany new commits dealing with things that came up along the way, which you can look at now - everything after `ImageWriter : Use ContextQuery instead of separate unpremult node` is new ( too bad github doesn't show them in order ).  I think you should now be able to resolve a lot of your specific comments, and then I can squash and get this looking tidier.

The main thing that would make your review easier that I didn't get to today is commenting on all your comments with a link to the commit that addresses them.  I sometimes do that, and it would make this a lot easier for you, but would also take a fair bit of time - up to you if you want to wait for me to do that.

johnhaddon commented on python/GafferImageUI/SelectViewUI.py:
```diff
@@ -0,0 +1,64 @@
+
+	"description",
+	"""
+	Pick one view from a multi-view image, resulting in a viewless image.  Has no effect if
```
I think we probably want to avoid the "viewless" terminology now? So maybe "outputting it as the default view"?

johnhaddon commented on src/GafferImage/SelectView.cpp:
```diff
@@ -0,0 +1,251 @@
+void SelectView::initViewScope( Context::EditableScope &scope, std::string &viewStorage ) const
+{
+	// TODO - consider API where we don't allocate scope when viewNames.size() == 0.
+	// Maybe should return boost::optional< EditableScope >?
```
Should we replace `initViewScope()` with `ViewScope::setViewNameChecked()` now?

johnhaddon reviewed (commented):
> OK, I think this is actually ready for a preliminary re-review

Thanks Daniel!

I've gone through all my previous comments and resolved the relevant ones. I've also looked through all the new commits (except ones marked FIX, which I assume won't contain stuff other than the fixes I saw by going through the comments). Made a few more comments inline in this review, and spotted a few other things while playing around with it :

- [x] Viewer is defaulting to the `left` view, whereas I would expect `default`.
- [x] `+` "nodule" for CreateViews doesn't work, presumably because the underlying plug is a NameValuePlug which won't accept drags from an ImagePlug. Will either need a custom gadget, or we can remove it for now (and rely on the button in the NodeEditor).
- [ ] We should update the `GafferImageTest::ContextSanitiser` to check that `image:view` isn't set when querying view names.
- [ ] The FormatQuery UI seems a little bit confusing now there's an input `view` plug immediately above the output plugs in the NodeEditor. I think this is exacerbated by the widget for the `format` plug looking very much like an input plug. Perhaps we can move all the output plugs into a collapsible `Out` section as we've done for some of the other queries.

Now seems like a good time to squash things down into whatever form you want them to be. I'm happy to review the outstanding fixes either as separate commits or within the squashes...

Cheers...
John

danieldresser-ie commented on python/GafferImageTest/ImageReaderTest.py:
```diff
@@ -478,7 +479,13 @@ def validateChannels( channels ):
 		validateChannels( [ ( "rgba.main.R", 0.1 ), ( "rgba.main.G", 0.2 ), ( "rgba.main.B", 0.3 ), ( "character.main.red", 0.4 ), ( "character.main.green", 0.5 ), ( "character.main.blue", 0.6 ) ] )
 		reader["channelInterpretation"].setValue( GafferImage.ImageReader.ChannelInterpretation.Specification )
-		validateChannels( [ ( "R", 0.1 ), ( "G", 0.2 ), ( "B", 0.3 ), ( "red", 0.4 ), ( "green", 0.5 ), ( "blue", 0.6 ) ] )
+		# This file has a view of "main" - by default, we assume that this is a Nuke file which should be
+		# interpreted as viewless, but if you ask for the specification interpretation, then you actually get
+		# a view
```
Addressed by adding the view parameter to validateChannels in e3d74e3

danieldresser-ie commented on python/GafferUI/PresetsPlugValueWidget.py:
```diff
@@ -79,16 +79,26 @@ def _updateFromPlugs( self ) :
+			except:
+				errored = True
+
+			isCustom = False
+			try:
+				allowCustom = sole( ( Gaffer.Metadata.value( p, "presetsPlugValueWidget:allowCustom" ) for p in self.getPlugs() ) )
+				isCustom = any( Gaffer.Metadata.value( p, "presetsPlugValueWidget:isCustom" ) for p in self.getPlugs() )
+				isCustom = allowCustom and ( isCustom or currentPreset == "" )
+			except:
+				pass
```
Yeah, good point.  I had set up a test script with an erroring plug, and was trying to get it to display, and went a bit far, and then missed the isCustom guard when removing the unnecessary stuff.

What I think would be good to deal with in the future is that having a visibility activator for one plug that depends on the value of another plug ( which is quite common ), can cause the whole UI to fail to draw if the plug referenced in the activator expression has an evaluation error.  Not something to deal with now though.

danieldresser-ie commented:
OK, should have rerun the non-image tests after that last round of changes, there are some test failures, but nothing that should prevent review, and all look like quick fixes tomorrow.

johnhaddon commented on src/GafferImage/CreateViews.cpp:
```diff
@@ -114,59 +113,145 @@ const Gaffer::ArrayPlug *CreateViews::viewsPlug() const
-	for( const GraphComponentPtr &p : viewsPlug()->children() )
+	if( output != indexPlug() )
+	{
+		ImageNode::hash( output, context, h );
```
We also need this call to the base class when we're hashing `indexPlug()`.

johnhaddon commented on src/GafferImage/CreateViews.cpp:
```diff
@@ -114,59 +113,145 @@ const Gaffer::ArrayPlug *CreateViews::viewsPlug() const
+	const std::string &currentView = context->get<std::string>( ImagePlug::viewNameContextName );
+
+	Context::EditableScope s( context );
+	s.remove( ImagePlug::viewNameContextName );
```
I think the call to `remove()` is invalidating the reference we take on L197.

johnhaddon replied:
Daniel pointed out that this is fine, because the removal is from a new EditableScope, not the original context.

johnhaddon commented on python/GafferImageTest/FormatQueryTest.py:
```diff
@@ -88,9 +90,56 @@ def test( self ) :
+		formatQuery["image"].setInput( views["out"] )
+
+		# The result when contextView and queryView are both not overridden is technically undefined behaviour
+		# - views has just left and right views, so it's illegal to ask for "default".  But we know in practice
+		# that CreateViews will currently return the default value of an ImagePlug, which has an HD format
```
This comment seems to contradict the `assertRaisesRegex()` at L127. Seems like the assertion is correct and the comment is wrong?

johnhaddon commented on python/GafferImageTest/ImageSamplerTest.py:
```diff
@@ -91,5 +93,52 @@ def testChannelsPlug( self ) :
+		sampler["pixel"].setValue( imath.V2f( 50.5 ) )
+
+		# The result when contextView and queryView are both not overridden is technically undefined behaviour
+		# - views has just left and right views, so it's illegal to ask for "default".  But we know in practice
+		# that CreateViews will currently return the default value of an ImagePlug, which is black
```
Same as for FormatQueryTest?

johnhaddon commented on python/GafferImageTest/ImageStatsTest.py:
```diff
@@ -241,6 +242,52 @@ def testRepeatedChannels( self ) :
+		s["in"].setInput( views["out"] )
+		s["area"].setValue( reader["out"]["format"].getValue().getDisplayWindow() )
+
+		# The result when contextView and queryView are both not overridden is technically undefined behaviour
+		# - views has just left and right views, so it's illegal to ask for "default".  But we know in practice
+		# that CreateViews will currently return the default value of an ImagePlug, which is black
```
Ditto?

johnhaddon reviewed (changes requested):
Thanks Daniel! I've been through all my old comments and read through all the latest "FIX" commits. Spotted a couple of minor corrections (see comments inline) but I think we're basically there. Would be great if you could get everything squashed and ready to go...

danieldresser-ie commented:
Had a last minute scare with the Arnold tests failing, and tracking the issue down to BleedFill - when I realized that the single view network was now failing, I was worried that I'd overlooked something fundamental that could break user networks.  But in the end, I think it's OK - the issue was the Loop in BleedFill that drives the number of iterations from the size of the image.  The viewNames plug is evaluated without viewName in the context, but evaluating viewNames on the Loop requires evaluating the number of iterations, which depends on the size of the image, which needs viewName in the context.  I don't think this sort of issue is likely to arise in user networks.

I've fixed it for now by using a FormatQuery hardcoded to the default view - long term, we should probably make BleedFill actually operate on multi-view images, but that might be easiest to do using CollectViews once it exists.

Other than that, I've cleaned up the couple of things you mentioned, cleaned up the GafferScene, GafferOSL, and GafferArnold tests, and rebased.  I was probably a bit sloppy with the Changelog, wasn't sure how much to try to put things in the right commit ... let me know if I should take another pass on cleaning that up.
